compiler/rustc_infer/src/infer/opaque_types.rs

```diff
@@ -645,3 +657,93 @@ fn may_define_opaque_type(tcx: TyCtxt<'_>, def_id: LocalDefId, opaque_hir_id: hi
     );
     res
 }
+
+/// Checks that the item may register hidden types for the
+/// opaque type, if the opaque type shows up in its signature.
+#[instrument(level = "debug", skip(tcx), ret)]
+pub fn may_define_impl_trait_in_assoc_ty_modulo_sig<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: LocalDefId,
+    opaque_def_id: LocalDefId,
+) -> Option<impl TypeVisitable<TyCtxt<'tcx>>> {
+    let sig = match tcx.def_kind(def_id) {
+        DefKind::AssocFn => Ok(tcx.fn_sig(def_id).skip_binder()),
+        DefKind::AssocConst | DefKind::AssocTy => Err(tcx.type_of(def_id).skip_binder()),
```

Comment:
Are you using `Ok` and `Err` as `Either::{Left, Right}`? If so, please define a new enum and derive `TypeVisitable` instead, the `None` + `Err` usage here confused me a bit :sweat_smile: